software/controller/lib/core/circular_buffer.h

```diff
@@ -26,37 +26,40 @@ limitations under the License.
 // and the interrupt handlers, so it needs to be thread safe.
 // I'm disabling interrupts during the critical sections to
 // ensure that's the case.
-template <class T, int N> class CircBuff {
-  volatile T buff[N];
-  volatile int head, tail;
+template <class T, uint N> class CircBuff {
+  // Uses an array of size N+1 to hold all N elements of the buffer, as
+  // buff_[head_] is by definition inaccessible.
+  // This also makes the template safe against N = 0.
+  volatile T buff_[N + 1];
+  volatile int head_, tail_;
 
 public:
-  CircBuff() { head = tail = 0; }
+  CircBuff() { head_ = tail_ = 0; }
 
   // Return number of elements available in the buffer to read.
-  int FullCt() {
+  int FullCt() const {
```

Comment:
Maybe these functions as well at the class can use full words? Like "count" and "circular buffer"?